scanner/indexer/remote.go

```diff
@@ -0,0 +1,66 @@
+package indexer
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/quay/claircore"
+	"github.com/stackrox/rox/generated/internalapi/scanner/v4"
+	"github.com/stackrox/rox/pkg/clientconn"
+	"github.com/stackrox/rox/pkg/mtls"
+	"github.com/stackrox/rox/scanner/mappers"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// remoteIndexer is the Indexer implementation that connects to a remote indexer
+// using gRPC.
+type remoteIndexer struct {
+	conn    *grpc.ClientConn
+	indexer v4.IndexerClient
+}
+
+// NewRemoteIndexer connect to the gRPC address and creates a new remote indexer.
+func NewRemoteIndexer(ctx context.Context, address string) (*remoteIndexer, error) {
+	connOpt := clientconn.Options{
+		TLS: clientconn.TLSConfigOptions{
+			GRPCOnly:           true,
+			InsecureSkipVerify: true,
+		},
+	}
+	// TODO: [ROX-19050] Set the Scanner V4 TLS validation and the correct subject
+	//       when certificates are ready.
+	conn, err := clientconn.GRPCConnection(ctx, mtls.ScannerSubject, address, connOpt)
+	if err != nil {
+		return nil, err
+	}
+	return &remoteIndexer{
+		conn:    conn,
+		indexer: v4.NewIndexerClient(conn),
+	}, nil
+}
+
+// Close closes the remote indexer.
+func (r *remoteIndexer) Close(ctx context.Context) error {
+	return r.conn.Close()
+}
+
+// GetIndexReport call the remote service to retrieve an IndexReport for the given hash ID.
+func (r *remoteIndexer) GetIndexReport(ctx context.Context, hashID string) (*claircore.IndexReport, bool, error) {
+	resp, err := r.indexer.GetIndexReport(ctx, &v4.GetIndexReportRequest{HashId: hashID})
+	if err != nil {
+		if e, ok := status.FromError(err); ok && e.Code() == codes.NotFound {
+			return nil, false, nil
+		}
+		return nil, false, err
+	}
+	if !resp.GetSuccess() {
+		return nil, true, fmt.Errorf("report failed: state %s: %s", resp.GetState(), resp.GetErr())
```

Comment:
this should return `false` I think, right?

Reply:
The report was found, but it was not successful. Maybe also return the `resp` here, but for some reason, I thought just the flag was enough.

Reply:
is it always true it was found?

Reply:
No, if it was not found in the DB, we return `false` in lines 60 and 62.

Reply:
sorry I still don't get why this is `true` then?

Reply:
It is true in line 65 because the report was found in the DB. The report status is not successful, and the State and Err will tell you why. I conveniently returned them in the error message, so I decided to not return the report back to the caller. But the report still exists in the DB.

Reply:
[PR was approved](https://github.com/stackrox/stackrox/pull/8579#pullrequestreview-1733147056), assuming this is resolved.